### PR TITLE
finish sign up

### PR DESCRIPTION
**Pull request by weilingooi**
finish sign up

### CODE REVIEW

**Comment by weilingooi** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -39,3 +38,76 @@ textarea{
...
+  margin-right: 10px;
+}
+
+.gravatar_edit {
+  margin-top: 15px;
+}
+
+/*forms */
+input, textarea, select, .uneditable-input{
+  border: 1px solid #bbb;
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]: please move color variable to other file, then call back here

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -39,3 +38,76 @@ textarea{
...
+
+.field_with_errors {
+  //@extend .has-error;
+  .form-control {
+  li {
+    color: red;
+  }
+}
+}
+}
```
Comment (chungpv-1008) [inline app/assets/stylesheets/custom.scss]: add end line (blank line) for each files

File: app/assets/stylesheets/users.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (chungpv-1008) [inline app/assets/stylesheets/users.scss]: remove this file, because not used to

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      if @user.save
+          flash[:success] = t ".welcome"
+          redirect_to@user
+      else
+        render :new
+      end
+  end
+  
+  private
+  def user_params
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]:
```suggestion
before_acton :load_user, only: :show

  def show; end

  def new
      @user = User.new
  end
  
  def create
      @user = User.new user_params
      
      if @user.save
          flash[:success] = t ".welcome"
          redirect_to@user
      else
        flash.now[:danger] = ...
        render :new
      end
  end
  
  private

  def load_user
     @user = User.find_by id: params[:id]
     return if @user.present?
     
     flash[:danger] = 
     redirect_to ...
  end

  def user_params
```

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      params.require(:user).permit(:name, :email, :password, :password_confirmation)
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]: in 'user_model' file, create constant USER_PARAMS = %i(name email ...).freeze
```suggestion
      params.require(:user).permit USER::USER_PARAMS
```

File: app/helpers/users_helper.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+module UsersHelper
+  def gravatar_for(user)
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    gravatar_url = "#{Settings.signup.gravatar.url}#{gravatar_id}"
+    image_tag(gravatar_url, alt: user.name, class:"gravatar")
+  end
```
Comment (chungpv-1008) [inline app/helpers/users_helper.rb]:
```suggestion
  def gravatar_for user
    gravatar_id = Digest::MD5::hexdigest user.email.downcase
    gravatar_url = "#{Settings.signup.gravatar.url}#{gravatar_id}"
    image_tag gravatar_url, alt: user.name, class:"gravatar"
  end
```

File: app/views/users/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,31 @@
+<% provide :title, t(".sign_up") %> 
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+	
+	  <% if @user.errors.any? %> 
+	    <div id="error_explanation">
+          <div class="alert alert-danger"> 
+            <%= t ".the_form_contains" %>
+            <%= pluralize @user.errors.count, "error" %>.
```
Comment (chungpv-1008) [inline app/views/users/new.html.erb]: please use 'size' instead of 'count'

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+      <%= f.label :password_confirmation, t(".confirmation") %>
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+      <%= f.submit t(".create_my_account"), class: "btn btn-primary" %>
+    <% end %>
+  </div>
+</div>
```
Comment (chungpv-1008) [inline app/views/users/new.html.erb:30]: add end line for this file

File: config/routes.rb [hunk 9]
```diff
@@ -3,4 +3,9 @@
   
   get "static_pages/home"
   get "static_pages/help"
+  get "users/show"
+  get "users/new"
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
  resources :users, only: %i(new create show)
```

File: config/routes.rb [hunk 10]
```diff
@@ -3,4 +3,9 @@
...
+  get "/signup", to: "users#new"
+  post "/signup", to: "users#create"
+  resources :users
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
```

File: app/assets/stylesheets/custom.scss (hunk 1 above)
Comment (weilingooi) [reply app/assets/stylesheets/custom.scss:82]: do not really understand the sentence, can I have further explanation?

File: app/assets/stylesheets/users.scss (hunk 3 above)
Comment (weilingooi) [reply app/assets/stylesheets/users.scss]: done

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/helpers/users_helper.rb (hunk 6 above)
Comment (weilingooi) [reply app/helpers/users_helper.rb]: done

File: app/views/users/new.html.erb (hunk 7 above)
Comment (weilingooi) [reply app/views/users/new.html.erb]: done

File: app/views/users/new.html.erb (hunk 8 above)
Comment (weilingooi) [reply app/views/users/new.html.erb:30]: done

File: config/routes.rb (hunk 9 above)
Comment (weilingooi) [reply config/routes.rb]: done

File: config/routes.rb (hunk 10 above)
Comment (weilingooi) [reply config/routes.rb]: done

File: app/assets/stylesheets/custom.scss (hunk 2 above)
Comment (weilingooi) [reply app/assets/stylesheets/custom.scss]: done

File: app/assets/stylesheets/custom.scss (hunk 1 above)
Comment (weilingooi) [reply app/assets/stylesheets/custom.scss:82]: > please move color variable to other file, then call back here
do not really understand the sentence, can I have further explanation?

Comment (weilingooi) [reply app/assets/stylesheets/custom.scss]: do not really understand the sentence, can I have further explanation?
> please move color variable to other file, then call back here

**Comment by weilingooi** [comment]
ready

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      @user = User.find_by id: params[:id]
+      return if @user.present?
+      
+      flash[:danger] = t ".user.user_not_found "
+      redirect_to root_path
+   end
+  
+  def user_params
+      params.require(:user).permit User::USERS_PARAMS
+  end
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: Notes: use only TWO spaces.

File: app/models/user.rb [hunk 12]
```diff
@@ -1,4 +1,5 @@
 class User < ApplicationRecord
+  USERS_PARAMS = %i(name email password password_confirmation).freeze
```
Comment (saobv-2223) [inline app/models/user.rb:2]: bellow constant variable, have one emtpy line

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,31 @@
+<% provide :title, t(".sign_up") %> 
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+	
+	  <% if @user.errors.any? %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]: remove the empty line
use 2 spaces

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -0,0 +1,35 @@
+class UsersController < ApplicationController
+before_action :load_user, only: :show
+
+  def show;end
```
Comment (LVQA1998) [inline app/controllers/users_controller.rb]:
```suggestion
  def show; end
```

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  def create
+      @user = User.new user_params
+      
+      if @user.save
+          flash[:success] = t ".welcome"
+          redirect_to@user
+      else
+        flash.now[:danger] = t ".error"
+        render :new
+      end
```
Comment (LVQA1998) [inline app/controllers/users_controller.rb]: ident 2 space

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  
+  private
+  
+  def load_user
+      @user = User.find_by id: params[:id]
+      return if @user.present?
+      
+      flash[:danger] = t ".user.user_not_found "
+      redirect_to root_path
+   end
```
Comment (LVQA1998) [inline app/controllers/users_controller.rb]: ident

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (LVQA1998) [inline app/controllers/users_controller.rb]: same

File: app/views/layouts/application.html.erb [hunk 17]
```diff
@@ -7,11 +7,15 @@
...
     <%= javascript_pack_tag "application", "data-turbolinks-track": "reload" %>
+	
   </head>
 
   <body>
   <%= render "layouts/header" %> 
   <div class="container">
+    <% flash.each do |message_type, message| %>
+	  <div class="alert alert-<%= message_type %>"><%= message %></div> 
+	<% end %>
```
Comment (LVQA1998) [inline app/views/layouts/application.html.erb]:
```suggestion
    <% flash.each do |message_type, message| %>
	<div class="alert alert-<%= message_type %>"><%= message %></div> 
    <% end %>
```

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/models/user.rb (hunk 12 above)
Comment (weilingooi) [reply app/models/user.rb:2]: done

File: app/views/users/new.html.erb (hunk 13 above)
Comment (weilingooi) [reply app/views/users/new.html.erb]: done

File: app/controllers/users_controller.rb (hunk 14 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 15 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 16 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/views/layouts/application.html.erb (hunk 17 above)
Comment (weilingooi) [reply app/views/layouts/application.html.erb]: comment
this is what I have on my laptop, I didn't change anything.I couldn't see what is wrong?

**Comment by weilingooi** [comment]
ready

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    @user = User.new user_params
+      
+      if @user.save
+        flash[:success] = t ".welcome"
+        redirect_to@user
+      else
+        flash.now[:danger] = t ".error"
+        render :new
+      end
+  end
```
Comment (LVQA1998) [inline app/controllers/users_controller.rb]:
```suggestion
  def create
    @user = User.new user_params
      
    if @user.save
      flash[:success] = t ".welcome"
      redirect_to @user
    else
      flash.now[:danger] = t ".error"
      render :new
    end
  end
```

File: app/views/layouts/application.html.erb [hunk 19]
```diff
@@ -7,11 +7,15 @@
 
     <%= stylesheet_link_tag "application", media: "all", "data-turbolinks-track": "reload" %>
     <%= javascript_pack_tag "application", "data-turbolinks-track": "reload" %>
+
```
Comment (LVQA1998) [inline app/views/layouts/application.html.erb]:
```suggestion
```

File: app/views/layouts/application.html.erb (hunk 17 above)
Comment (LVQA1998) [reply app/views/layouts/application.html.erb]: You can use text editor to check it

File: app/views/users/new.html.erb [hunk 20]
```diff
@@ -0,0 +1,30 @@
+<% provide :title, t(".sign_up") %> 
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+	  <% if @user.errors.any? %>
```
Comment (LVQA1998) [inline app/views/users/new.html.erb]:
```suggestion
	<% if @user.errors.any? %> 
```

File: app/assets/stylesheets/custom.scss [hunk 21]
```diff
@@ -39,3 +38,77 @@ textarea{
...
+
+.field_with_errors {
+  //@extend .has-error;
+  .form-control {
+  li {
+    color: red;
+
+  }
+}
+}
```
Comment (manhnt-0960) [inline app/assets/stylesheets/custom.scss]: please check indent again

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+  def create
+    @user = User.new user_params
+      
+      if @user.save
+        flash[:success] = t ".welcome"
+        redirect_to@user
+      else
+        flash.now[:danger] = t ".error"
+        render :new
+      end
```
Comment (manhnt-0960) [inline app/controllers/users_controller.rb]: don't need indent

File: app/controllers/users_controller.rb (hunk 18 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/views/layouts/application.html.erb (hunk 19 above)
Comment (weilingooi) [reply app/views/layouts/application.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 17 above)
Comment (weilingooi) [reply app/views/layouts/application.html.erb]: done

File: app/views/users/new.html.erb (hunk 20 above)
Comment (weilingooi) [reply app/views/users/new.html.erb]: done

**Comment by weilingooi** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 23]
```diff
@@ -39,3 +38,76 @@ textarea{
...
+
+.field_with_errors {
+  //@extend .has-error;
+  .form-control {
+    li {
+    color: red;
+     }
+  } 
+}
+}
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: check indent again, I see redundant } in line 113

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -0,0 +1,35 @@
+class UsersController < ApplicationController
+before_action :load_user, only: :show
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: add 2 spaces to this line

File: app/views/layouts/application.html.erb [hunk 25]
```diff
@@ -12,8 +12,11 @@
...
   <div class="container">
+    <% flash.each do |message_type, message| %>
+      <div class="alert alert-<%= message_type %>"><%= message %></div> 
+    <% end %>
     <%= yield %> 
   </div>
 	
   </body>
-</html>
+</html>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]: Please, add an empty line

File: app/controllers/users_controller.rb (hunk 24 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/views/layouts/application.html.erb (hunk 25 above)
Comment (weilingooi) [reply app/views/layouts/application.html.erb]: done

**Comment by weilingooi** [comment]
ready

File: app/views/layouts/application.html.erb (hunk 25 above)
Comment (saobv-2223) [reply app/views/layouts/application.html.erb]: You have not yet fixed this mistake, add an empty line after row 23

File: app/views/users/new.html.erb [hunk 26]
```diff
@@ -0,0 +1,30 @@
+<% provide :title, t(".sign_up") %> 
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
+      <% if @user.errors.any? %> 
+	    <div id="error_explanation">
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]:
```suggestion
	  <div id="error_explanation">
```

File: app/views/layouts/application.html.erb (hunk 25 above)
Comment (weilingooi) [reply app/views/layouts/application.html.erb]: done

File: app/views/users/new.html.erb (hunk 26 above)
Comment (weilingooi) [reply app/views/users/new.html.erb]: done

**Comment by weilingooi** [comment]
ready

File: app/controllers/users_controller.rb [hunk 27]
```diff
@@ -0,0 +1,35 @@
+class UsersController < ApplicationController
+
+  def show; end
+  before_action :load_user, only: :show
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
  before_action :load_user, only: :show
  
  def show; end
```

File: app/views/users/new.html.erb [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+          <div class="alert alert-danger"> 
+            <%= t ".the_form_contains" %>
+            <%= pluralize @user.errors.size, "error" %>.
+          </div>
+          <ul>
+          <% @user.errors.full_messages.each do |msg| %>
+            <li><%= msg %></li>
+          <% end %>
+          </ul>
+        </div>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]:
```suggestion
        <div class="alert alert-danger"> 
          <%= t ".the_form_contains" %>
          <%= pluralize @user.errors.size, "error" %>.
        </div>
        <ul>
          <% @user.errors.full_messages.each do |msg| %>
            <li><%= msg %></li>
          <% end %>
        </ul>
      </div>
```

File: app/views/users/new.html.erb [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+      <%= f.label :password_confirmation, t(".confirmation") %>
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+      <%= f.submit t(".create_my_account"), class: "btn btn-primary" %>
+    <% end %>
+  </div>
+</div>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb:30]: add endline, please

File: app/controllers/users_controller.rb (hunk 27 above)
Comment (weilingooi) [reply app/controllers/users_controller.rb]: done

File: app/views/users/new.html.erb (hunk 28 above)
Comment (weilingooi) [reply app/views/users/new.html.erb]: done

File: app/views/users/new.html.erb (hunk 29 above)
Comment (weilingooi) [reply app/views/users/new.html.erb:30]: done

**Comment by weilingooi** [comment]
ready

File: app/views/users/new.html.erb [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+          <%= t ".the_form_contains" %>
+          <%= pluralize @user.errors.size, "error" %>.
+        </div>
+        <ul>
+          <% @user.errors.full_messages.each do |msg| %>
+            <li><%= msg %></li>
+          <% end %>
+        </ul>
+      </div>
+      <% end %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]:
```suggestion
      <% if @user.errors.any? %> 
	  <div id="error_explanation">
          <div class="alert alert-danger"> 
            <%= t ".the_form_contains" %>
            <%= pluralize @user.errors.size, "error" %>.
          </div>
          <ul>
              <% @user.errors.full_messages.each do |msg| %>
              <li><%= msg %></li>
            <% end %>
          </ul>
        </div>
      <% end %>
```

Comment (weilingooi) [reply app/views/users/new.html.erb]: done

**Comment by weilingooi** [comment]
ready

**Pull request merged**